### PR TITLE
Scope Helix Job Monitor to the current stage attempt

### PR DESCRIPTION
**Pull request by mmitche**
## Purpose

Update the Helix Job Monitor so it **gates completion on the current stage attempt while resubmitting a previous attempt''s unfinished/failed work into the current one**, because today it re-discovers a previous attempt''s Helix jobs and can hang forever on work items permanently stranded in `Waiting` (e.g. after a queue purge) — never terminating and re-burning its full timeout on every retry (#17156).

Scoping strictly to the current attempt is not enough on its own: Azure DevOps'' *retry-failed-jobs* gesture re-runs only failed jobs, so if the Helix submitters passed and only the monitor timed out, the current attempt has no Helix work — and the monitor must still reconcile (upload results, resubmit failures) the previous attempt''s work rather than exit 0 and discard it.

## What changed

- Monitor tracks a new `--stage-attempt` (`SYSTEM_STAGEATTEMPT`); completion gating considers only current-attempt Helix jobs.
- Previous-attempt work is reconciled per logical work stream (AzDO job + queue): resubmit failed/unfinished items into the current attempt (stamped with the current attempt so the monitor gates on its own resubmission), upload passed results, and fail fast when work can never be resubmitted (e.g. purged queue).
- Six pipeline-emulating `AttemptScoped_*` tests cover the corner cases; all 179 `Microsoft.DotNet.Helix.Sdk.Tests` pass. `JobMonitorRunner.Design.md` updated.

Contributes to #17156.

### CODE REVIEW

**Review by premun — DISMISSED** [review]
@mmitche this only solves when we re-run the whole stage but it does not solve when we time out and the helix jobs from the same attempt will never be completed, right? For that we need to resubmit all jobs that are incomplete when we're re-running the monitor only.

**Comment by mmitche** [comment]
@premun I haven't looked at this yet, so not sure. On my todo list today.

**Comment by mmitche** [comment]
Updated this PR after review feedback: the earlier "scope strictly to the current attempt and ignore previous-attempt jobs" design was too naive.

**Why it was wrong:** Azure DevOps'' *retry-failed-jobs* gesture re-runs only failed jobs. If the Helix submitters passed and only the monitor timed out, the submitters do not re-run, so the current attempt has zero Helix work — the naive monitor would exit 0 immediately and silently discard the previous attempt''s results and failures. There were also several related corner cases (resubmissions stamped with the wrong attempt, cancel-mid-retry, still-running previous incarnations during a fast rerun, unlinked rerun duplicates, and un-resubmittable/purged-queue work).

**New model:** gate completion on the current attempt, but reconcile previous-attempt work into it by resubmission, decided per logical work stream (submitter + queue) rather than per attempt. Each of the six corner cases is now pinned by a pipeline-emulating `AttemptScoped_*` test (each mutation-verified to fail if its mechanism is reverted). Full details are in the updated PR description and `JobMonitorRunner.Design.md` §2.1/§2.3/§2.3.1.

File: src/Microsoft.DotNet.Helix/JobMonitor/JobMonitorRunner.cs [hunk 1]
```diff
@@ -182,22 +206,35 @@ private async Task<IReadOnlyList<HelixJobInfo>> ExecuteRetryPassAsync(Cancellati
                         .DistinctBy(wi => wi.Name, StringComparer.OrdinalIgnoreCase)
                 ];
 
-                HelixJobInfo resubmitted = await _helix.ResubmitWorkItemsAsync(completedJob, failedWorkItems, cancellationToken);
+                HelixJobInfo resubmitted = await _helix.ResubmitWorkItemsAsync(
+                    latest, failedWorkItems, _options.StageAttempt, cancellationToken);
                 if (resubmitted is null)
                 {
+                    // Previous-attempt work that can never run again (e.g. its queue was removed)
```
Comment (premun) [inline src/Microsoft.DotNet.Helix/JobMonitor/JobMonitorRunner.cs:213]:
```suggestion
                    // Previous-attempt work that can never run again (e.g. its queue was purged)
```

**Review by premun — APPROVED** [review]
I went through it and I think it does make sense like this
